### ISSUE TITLE
unused-type-ignore

### ISSUE THREAD

**Issue by janosh** [astral-sh/ruff#3081]
Very tall ask considering I don't see how to do this without depending on `mypy` but wanted to at least float the suggestion.

Similar to how useless `# noqa` directives are removed by [`RUF100`](https://beta.ruff.rs/docs/rules/#ruff-specific-rules-ruf), it would be cool to auto-remove `# type: ignore` on lines that no longer issue a `mypy` error.

**Comment by charliermarsh** [comment]
Haha yeah we can't support this right now. But I'd like to support it in the future :eyes: I'll leave it open for now!

**Comment by stinodego** [comment]
@janosh In case you didn't know, `mypy` can at least warn you when this happens (but doesn't autofix it for you):
https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-warn-unused-ignores

**Comment by janosh** [comment]
@stinodego Thanks for bringing that up, should have mentioned it myself. I'm already using that flag. `mypy` is just painfully slow so if `ruff` could do it, would be a big QoL improvement.

**Comment by twoertwein** [comment]
While I would like this feature, I think it is out of scope for ruff:

- per [PEP 484](https://peps.python.org/pep-0484/#compatibility-with-other-uses-of-function-annotations) `typing: ignore` is not mypy-specific but intended for all type checkers (I assume ruff doesn't want to implement all type checkers)
- mypy/pyright/... and the python type system are still changing, does ruff really want to keep up with this (sounds like a lot of work for little benefit given that type checkers themselves have this check already)? It would also mean having to know the version of mypy and reading its config file to know which rules are used.

**Comment by Avasam** [comment]
> Haha yeah we can't support this right now. But I'd like to support it in the future 👀 I'll leave it open for now!

Wouldn't this now fully land on ty's side? ty as a replacement for mypy means this is completely obsolete (and even then I'm not sure ruff/ty should try to guess what's unused for mypy)

**Comment by MichaReiser** [comment]
I agree; I don't think it makes sense for Ruff to implement such a rule. It would require exact mypy (or any other type checker) parity, which doesn't make sense for Ruff. ty has such a rule; mypy also has a flag. If performance is a concern, multiple faster type checkers are available now.
